pkg/reconciler/iploop.go

```diff
@@ -0,0 +1,162 @@
+package reconciler
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dougbtv/whereabouts/pkg/allocate"
+	"github.com/dougbtv/whereabouts/pkg/logging"
+	"github.com/dougbtv/whereabouts/pkg/storage"
+	"github.com/dougbtv/whereabouts/pkg/storage/kubernetes"
+	"github.com/dougbtv/whereabouts/pkg/types"
+)
+
+type ReconcileLooper struct {
+	cancelFunc  func()
+	ctx         context.Context
+	k8sClient   kubernetes.Client
+	livePodRefs []string
+	orphanedIPs []OrphanedIPReservations
+}
+
+type OrphanedIPReservations struct {
+	Pool        storage.IPPool
+	Allocations []types.IPReservation
+}
+
+func NewReconcileLooper(kubeConfigPath string) (*ReconcileLooper, error) {
+	logging.Debugf("NewReconcileLooper - Kubernetes config file located at: %s", kubeConfigPath)
+	k8sClient, err := kubernetes.NewClient(kubeConfigPath)
+	if err != nil {
+		return nil, logging.Errorf("failed to instantiate the Kubernetes client: %+v", err)
+	}
+	logging.Debugf("successfully read the kubernetes configuration file located at: %s", kubeConfigPath)
+
+	ctx, cancel := context.WithTimeout(context.Background(), storage.RequestTimeout)
+	podRefs, err := getPodRefs(*k8sClient)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	looper := &ReconcileLooper{
+		cancelFunc:  cancel,
+		ctx:         ctx,
+		k8sClient:   *k8sClient,
+		livePodRefs: podRefs,
+	}
+
+	if err := looper.findOrphanedIPsPerPool(); err != nil {
+		return nil, err
+	}
+	return looper, nil
+}
+
+func getPodRefs(k8sClient kubernetes.Client) ([]string, error) {
+	pods, err := k8sClient.ListPods()
+	if err != nil {
+		return nil, err
+	}
+
+	var podRefs []string
+	for _, pod := range pods {
+		podRefs = append(podRefs, fmt.Sprintf("%s/%s", pod.GetNamespace(), pod.GetName()))
+	}
+	return podRefs, err
+}
+
+func (rl *ReconcileLooper) findOrphanedIPsPerPool() error {
+	ipPools, err := rl.k8sClient.ListIPPools(rl.ctx)
+	if err != nil {
+		return logging.Errorf("failed to retrieve all IP pools: %v", err)
+	}
+
+	for _, pool := range ipPools {
+		orphanIP := OrphanedIPReservations{
+			Pool: pool,
+		}
+		for _, allocation := range pool.Allocations() {
+			logging.Debugf("the IP reservation: %s", allocation)
+			if allocation.PodRef == "" {
+				_ = logging.Errorf("pod ref missing for Allocations: %s", allocation)
+				continue
+			}
+			if !rl.isPodAlive(allocation.PodRef) {
+				logging.Debugf("pod ref %s is not listed in the live pods list", allocation.PodRef)
+				orphanIP.Allocations = append(orphanIP.Allocations, allocation)
+			}
+		}
+		if len(orphanIP.Allocations) > 0 {
+			rl.orphanedIPs = append(rl.orphanedIPs, orphanIP)
+		}
+	}
+
+	return nil
+}
+
+func (rl ReconcileLooper) isPodAlive(podRef string) bool {
+	for _, livePodRef := range rl.livePodRefs {
+		if podRef == livePodRef {
```

Comment:
Need to check whether the running Pod really owns the reserved IP.

StatefulSet Pods keep their names whenever they recreated, so it might happen that a StatefulSet Pod got new IP in the meantime, while previous IP is still reserved in IPPool pointing to the same running Pod, so the previous IP is unused, bot won't be released.